### PR TITLE
8357268: Use JavaNioAccess.getBufferAddress rather than DirectBuffer.address()

### PR DESCRIPTION
**Pull request by minborg**
This PR proposes to use  JavaNioAccess::getBufferAdress rather than DirectBuffer::address so that Buffer instances backed by MemorySegment instances can be used in classes that were not covered by https://github.com/openjdk/jdk/pull/25321

This PR passes tier1, tier2, and tier3 tests on multiple platforms and configurations.

### CODE REVIEW

File: src/java.base/windows/classes/sun/nio/ch/WindowsAsynchronousFileChannelImpl.java [hunk 1]
```diff
@@ -446,13 +448,15 @@ public void run() {
...
+                address = NIO_ACCESS.getBufferAddress(dst) + pos;
             } else {
+                // buf can never be backed by a memory segment
                 buf = Util.getTemporaryDirectBuffer(rem);
-                address = ((DirectBuffer)buf).address();
+                address = NIO_ACCESS.getBufferAddress(buf);
             }
 
             boolean pending = false;
+            NIO_ACCESS.acquireSession(buf);
```
Comment (minborg) [inline src/java.base/windows/classes/sun/nio/ch/WindowsAsynchronousFileChannelImpl.java:459]: Here, we acquire the session *after* we have obtained the address. This is safe as we do not touch the segment before it is acquired. If a segment is deallocated before we try to acquire the session, an exception will be thrown.

**Comment by AlanBateman** [comment]
We need to audit the tests for the APIs that take a byte buffer to ensure that we have tests to exercise these APIs with buffers that are views on a memory segment. That would help identify any bugs. 

The temporary buffer cache is internal so I think better to separate from this JBS issue and PR as these cases cannot be views on a memory segment. Future work to re-implement the temporary buffer cache can re-visit this.

**Comment by AlanBateman** [comment]
A question for folks onsecurity-dev. Are there tests for Cipher.doFinal, CipherSpi.engineUpdate, etc. that exercises cases where the ByteBuffer obtained from a memory segment?

File: src/java.base/share/classes/sun/nio/ch/Util.java [hunk 2]
```diff
@@ -43,7 +43,7 @@
 
 public class Util {
 
-    private static final JavaNioAccess NIO_ACCESS = SharedSecrets.getJavaNioAccess();
+    static final JavaNioAccess NIO_ACCESS = SharedSecrets.getJavaNioAccess();
```
Comment (AlanBateman) [inline src/java.base/share/classes/sun/nio/ch/Util.java:46]: There are methods in IOUtil for the classes in sun.nio.ch. So no need to expose NIO_ACCESS here.

File: src/java.base/windows/classes/sun/nio/ch/WindowsAsynchronousFileChannelImpl.java [hunk 3]
```diff
@@ -662,6 +668,7 @@ public void run() {
                     ioCache.remove(overlapped);
 
             } finally {
+                NIO_ACCESS.releaseSession(buf);
```
Comment (AlanBateman) [inline src/java.base/windows/classes/sun/nio/ch/WindowsAsynchronousFileChannelImpl.java:671]: I thin the change to WindowsAsynchronousFileChannel will need more than one reviewer as the I/O operation may complete on a different thread to the one that it was initiated on.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11AEADCipher.java [hunk 4]
```diff
@@ -759,8 +760,8 @@ private int implDoFinal(ByteBuffer inBuffer, ByteBuffer outBuffer)
                     long outAddr = 0;
                     byte[] outArray = null;
                     int outOfs = 0;
-                    if (outBuffer instanceof DirectBuffer dOutBuffer) {
-                        outAddr = dOutBuffer.address();
+                    if (outBuffer instanceof DirectBuffer) {
+                        outAddr = NIO_ACCESS.getBufferAddress(outBuffer);
```
Comment (AlanBateman) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11AEADCipher.java:764]: This looks right.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Cipher.java [hunk 5]
```diff
@@ -1012,8 +1013,8 @@ private int implDoFinal(ByteBuffer outBuffer)
                 long outAddr = 0;
                 byte[] outArray = null;
                 int outOfs = 0;
-                if (outBuffer instanceof DirectBuffer dOutBuffer) {
-                    outAddr = dOutBuffer.address();
+                if (outBuffer instanceof DirectBuffer) {
+                    outAddr = NIO_ACCESS.getBufferAddress(outBuffer);
```
Comment (AlanBateman) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Cipher.java:1017]: This looks right.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Digest.java [hunk 6]
```diff
@@ -289,7 +289,7 @@ protected void engineUpdate(ByteBuffer byteBuffer) {
             }
             NIO_ACCESS.acquireSession(byteBuffer);
             try {
-                token.p11.C_DigestUpdate(session.id(), dByteBuffer.address() + ofs, null, 0, len);
+                token.p11.C_DigestUpdate(session.id(), NIO_ACCESS.getBufferAddress(byteBuffer) + ofs, null, 0, len);
```
Comment (AlanBateman) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Digest.java:292]: `long address = NIO_ACCESS.getBufferAddress(byteBuffer);` might be clearer here.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Mac.java [hunk 7]
```diff
@@ -279,14 +279,14 @@ protected void engineUpdate(ByteBuffer byteBuffer) {
...
-            if (!(byteBuffer instanceof DirectBuffer dByteBuffer)) {
+            if (!(byteBuffer instanceof DirectBuffer)) {
                 super.engineUpdate(byteBuffer);
                 return;
             }
             int ofs = byteBuffer.position();
             NIO_ACCESS.acquireSession(byteBuffer);
             try  {
-                token.p11.C_SignUpdate(session.id(), dByteBuffer.address() + ofs, null, 0, len);
+                token.p11.C_SignUpdate(session.id(), NIO_ACCESS.getBufferAddress(byteBuffer) + ofs, null, 0, len);
```
Comment (AlanBateman) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Mac.java:289]: I think I would introduce a long address here too, only to make the line easier to read.

File: src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java [hunk 8]
```diff
@@ -947,8 +948,8 @@ ByteBuffer overlapDetection(ByteBuffer src, ByteBuffer dst) {
                 // side, we are not in overlap.
                 // NOTE: inPlaceArray does not apply here as direct buffers run
                 // through a byte[] to get to the combined intrinsic
-                if (((DirectBuffer) src).address() - srcaddr + src.position() >=
-                    ((DirectBuffer) dst).address() - dstaddr + dst.position()) {
+                if (NIO_ACCESS.getBufferAddress(src) - srcaddr + src.position() >=
+                        NIO_ACCESS.getBufferAddress(dst) - dstaddr + dst.position()) {
```
Comment (wenshao) [inline src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java:952]:
```suggestion
                if (NIO_ACCESS.getBufferAddress(src) - srcaddr + src.position() >=
                    NIO_ACCESS.getBufferAddress(dst) - dstaddr + dst.position()) {
```
Here you can use the same alignment style as before

File: src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java [hunk 9]
```diff
@@ -1602,7 +1603,7 @@ public int doFinal(ByteBuffer src, ByteBuffer dst)
                         NIO_ACCESS.acquireSession(dst);
                         try {
                             Unsafe.getUnsafe().setMemory(
-                                ((DirectBuffer)dst).address(),
+                                    NIO_ACCESS.getBufferAddress(dst),
                                 len + dst.position(), (byte) 0);
```
Comment (wenshao) [inline src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java:1607]:
```suggestion
                                NIO_ACCESS.getBufferAddress(dst),
                                len + dst.position(), (byte) 0);
```
Alignment

File: test/jdk/java/nio/channels/AsynchronousFileChannel/Basic.java [hunk 10]
```diff
@@ -563,15 +565,22 @@ static void testTruncate(Path file) throws IOException {
...
+            }
+            case 1 -> ByteBuffer.wrap(buf);
+            case 2 -> {
+                ByteBuffer bb = Arena.ofAuto().allocate(buf.length).asByteBuffer();
+                bb.put(buf);
+                bb.flip();
+                yield bb;
+            }
+            default -> throw new InternalError("Should not reach here");
+        };
```
Comment (wenshao) [inline test/jdk/java/nio/channels/AsynchronousFileChannel/Basic.java:583]:
```suggestion
        return switch (rand.nextInt(3)) {
            case 0  -> ByteBuffer.allocateDirect(buf.length)
                                 .put(buf)
                                 .flip();
            case 1  -> ByteBuffer.wrap(buf);
            case 2  -> Arena.ofAuto()
                            .allocate(buf.length)
                            .asByteBuffer()
                            .put(buf)
                            .flip();
            default -> throw new InternalError("Should not reach here");
        };
```
ByteBuffer supports chain programming style, so we can simplify it to this

File: src/java.base/share/classes/sun/nio/ch/Util.java [hunk 11]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2000, 2024, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2000, 2025, Oracle and/or its affiliates. All rights reserved.
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/nio/ch/Util.java:2]: I didn't see src changes in this class?

File: src/java.base/unix/classes/sun/nio/fs/UnixUserDefinedFileAttributeView.java [hunk 12]
```diff
@@ -166,14 +166,14 @@ public int read(String name, ByteBuffer dst) throws IOException {
...
+        if (dst instanceof sun.nio.ch.DirectBuffer) {
             NIO_ACCESS.acquireSession(dst);
             try {
-                long address = ddst.address() + pos;
+                long address = NIO_ACCESS.getBufferAddress(dst) + pos;
                 int n = read(name, address, rem);
                 dst.position(pos + n);
                 return n;
-            } finally {
+            }  finally {
```
Comment (valeriepeng) [inline src/java.base/unix/classes/sun/nio/fs/UnixUserDefinedFileAttributeView.java:176]: nit: extra space before finally?

**Comment by valeriepeng** [comment]
> A question for folks on security-dev. Are there tests for Cipher.doFinal, CipherSpi.engineUpdate, etc. that exercises cases where the ByteBuffer obtained from a memory segment?

I don't find any. We'd have to update them to cover the memory segment usage. Judging from the changed sources, general Cipher and SunPKCS11-specific tests update would be needed.

**Comment by AlanBateman** [comment]
> > A question for folks on security-dev. Are there tests for Cipher.doFinal, CipherSpi.engineUpdate, etc. that exercises cases where the ByteBuffer obtained from a memory segment?
> 
> I don't find any. We'd have to update them to cover the memory segment usage. Judging from the changed sources, general Cipher and SunPKCS11-specific tests update would be needed.

Thanks for checking. We need to create an issue in JBS for this. It may be a corner case to use these APIs and SPI with buffers that are views on a memory segment but it will happen at some point and would be good to ensure that it is covered by tests.

File: src/java.base/windows/classes/sun/nio/ch/WindowsAsynchronousFileChannelImpl.java (hunk 1 above)
Comment (valeriepeng) [reply src/java.base/windows/classes/sun/nio/ch/WindowsAsynchronousFileChannelImpl.java]: Is there documentation on when sessions should be acquired/released? Is this only for when using MemorySegment? In security area, the bytes are passed to the JNI code which calls the native library to process the bytes and I wonder if sessions should be acquired/released for these.

Comment (valeriepeng) [reply src/java.base/windows/classes/sun/nio/ch/WindowsAsynchronousFileChannelImpl.java]: Never mind, I see that there are already acquireSession/releaseSession calls when the ByteBuffer objects are accessed.

**Review by valeriepeng — APPROVED** [review]
Security source changes look fine.

File: src/java.base/unix/classes/sun/nio/fs/UnixUserDefinedFileAttributeView.java [hunk 13]
```diff
@@ -166,10 +166,10 @@ public int read(String name, ByteBuffer dst) throws IOException {
         assert (pos <= lim);
         int rem = (pos <= lim ? lim - pos : 0);
 
-        if (dst instanceof sun.nio.ch.DirectBuffer ddst) {
+        if (dst instanceof sun.nio.ch.DirectBuffer) {
```
Comment (AlanBateman) [inline src/java.base/unix/classes/sun/nio/fs/UnixUserDefinedFileAttributeView.java:169]: I assume this can be changed to test dst.isDirect() now and avoid the explicit cross package dependency.

File: src/java.base/windows/classes/sun/nio/ch/WindowsAsynchronousFileChannelImpl.java [hunk 14]
```diff
@@ -662,6 +666,7 @@ public void run() {
                     ioCache.remove(overlapped);
 
             } finally {
+                IOUtil.releaseScope(buf);
```
[the remark this answers is not in the document]
Comment (minborg) [reply src/java.base/windows/classes/sun/nio/ch/WindowsAsynchronousFileChannelImpl.java:669]: Sounds like a good idea: https://bugs.openjdk.org/browse/JDK-8357847

**Review by AlanBateman — APPROVED** [review]

**Review by AlanBateman — APPROVED** [review]
